### PR TITLE
change mezz_io register to support almazny

### PR DESCRIPTION
**Pull request by ciciwu213**
change pin numbers, set NOE of shift registers on Almazny to 0 all the time, remove some pin-out configuration to leave more space.

### CODE REVIEW

File: mirny.py [hunk 1]
```diff
@@ -321,13 +321,17 @@ def __init__(self, platform):
             tsi = TSTriple()
             self.specials += tsi.get_tristate(m)
             self.comb += [
-                tsi.o.eq(regs[3].write[i] | (0 if i >= 4 else
-                    (regs[1].write[11] & eem[i + 4].i))),
```
Comment (jordens) [inline mirny.py:325]: Why this change?

File: mirny.py [hunk 2]
```diff
@@ -321,13 +321,17 @@ def __init__(self, platform):
...
+                tsi.o.eq(regs[3].write[i]),
                 regs[3].read[i].eq(tsi.i),
                 tsi.oe.eq(regs[3].write[i + 8]),
                 regs[3].read[i + 8].eq(tsi.oe),
             ]
 
+        # set remaining Almazny shift register NOE to 0 at all times
+        for pin in platform.request("almazny_noe"):
+            x = Signal()
+            self.comb += [ x.eq(0), pin.eq(x) ]
```
Comment (jordens) [inline mirny.py]: just `pin.eq(0)`

File: mirny_cpld.py [hunk 3]
```diff
@@ -17,7 +17,11 @@
         # high: type 2 receiver, failsafe low
         ("fsen", 0, Pins("P80")),
 
-        ("mezz_io", 0, Pins("P57 P58 P59 P60 P61 P64 P68 P69")),
+        #almazny            CLK MOSI LT1 LT2  LT3 LT4 CLR
+        #                   IO8  11  10  13   15  0     3
+        ("mezz_io", 0, Pins("P94 P97 P96 P100 P92 P57 P60")),
```
Comment (jordens) [inline mirny_cpld.py]: This doesn't match the schematic in any way.

File: mirny_cpld.py [hunk 4]
```diff
@@ -17,7 +17,11 @@
...
+        #almazny                NOE1 NOE2 NOE3  4
+        ("almazny_noe", 0, Pins("P95 P98 P101 P58")),
```
Comment (jordens) [inline mirny_cpld.py:27]: The comment is incomplete don't reflect the schematics and the changes are needlessly forward-incompatible.

**Review by jordens — CHANGES_REQUESTED** [review]
* What's the analysis of backwards compatibility? Not possible or not attempted?
* Is this forward compatible with other mezzanines?

**Comment by sbourdeauducq** [comment]
> * What's the analysis of backwards compatibility? Not possible or not attempted?
> 
> * Is this forward compatible with other mezzanines?

The CPLD is full (isn't this why you routed only the first 8 pins? unfortunately it's not those that Almazny uses). So there isn't much wiggle room here. What are the other mezzanines?

**Comment by jordens** [comment]
There are 8 signals because Mirny v1.0 has 8 pins. See the respective issue. Unfortunately the Almazny design wasn't reviewed with respect to gateware/firmware. Since the current Almazny revision is also mechanically impossible to use, this is the time to make the electrical interface sensible. Given the current design there doesn't even appear to be a need to more than 8 pins.
I don't think the CPLD is full. What have you tried? Even if it were, there are parts of the logic, like the test points, that can be dropped without affecting anyone.
The other mezzanines include for example a set of RF high isolation switches that I've used.
A couple more problematic points with this PR beyond the ones already mentioned:

* What prevents you from allocating one or all of the remaining SPI address prefixes? That would come with a lot less baggage on multiple fronts.
* What's with the remaining MEZZ_IO signals?

**Comment by sbourdeauducq** [comment]
> Since the current Almazny revision is also mechanically impossible to use, this is the time to make the electrical interface sensible.

We have 3 Almazny boards already that we need to get to work; the solution for them is to remove the digital I/O connector, solder it on the other side of the PCB, and use a IDC cable to connect it to Mirny.
Considering the turnaround time of Almazny (they took over 8 months, and I gave up waiting for Creotech front panels and made my own) I'd rather not wait for a PCB respin.
I'm fine with flashing modified CPLD gateware for supporting these particular boards, in that case we should just make sure this stays compatible with the final version from the Kasli point-of-view.

File: mirny.py (hunk 2 above)
Comment (ciciwu213) [reply mirny.py]: It can't work with ``pin.eq(0)``, there is an error ``ERROR:HDLCompilers:168 - "mirny.v" line 313 No signals referenced in statement with implicit sensitivity list``.

**Comment by ciciwu213** [comment]
> What prevents you from allocating one or all of the remaining SPI address prefixes? That would come with a lot less baggage on multiple fronts.
What's with the remaining MEZZ_IO signals?

The main reason are cpld is almost full and we don't have much space for required changes to make almazny work. Even if we delete test points, eem to mezz if we add one more register, it cannot work as macrocells which design requires exceed device limit. So maybe we will use this version for only almazny support and use the original version for other purposes.

**Comment by jordens** [comment]
As explained above I don't think you are correct. The CPLD isn't full. I tried the suggestion above and it builds just fine.
With that in mind and and the quality of your approach I'm not going to merge this for now since it's highly unlikely to be a viable solution.
I would recommend taking a hard and comprehensive look at the suggestions above and a redesign for Almazny v1.1.
Surely, for the boards you have, the hack will suffice as a temporary fix.

File: Makefile [hunk 5]
```diff
@@ -11,6 +11,10 @@ build: build/mirny.vm6
 build/mirny.vm6: mirny.py mirny_cpld.py
 	python mirny_impl.py
 
+.PHONY: legacy-almazny
+legacy-almazny: mirny.py mirny_cpld.py
+	python mirny_impl.py --legacy-almazny
```
Comment (sbourdeauducq) [inline Makefile:16]: Name the output differently so it doesn't have to be a phony target?

File: mirny.py [hunk 6]
```diff
@@ -174,10 +174,10 @@ class Mirny(Module):
...
     | EEM 2         | MISO, MUXOUT           |
     | EEM 3         | CS                     |
-    | EEM 4         | SW0, MUXOUT0           |
-    | EEM 5         | SW1, MUXOUT1           |
-    | EEM 6         | SW2, MUXOUT2           |
-    | EEM 7         | SW3, MUXOUT3           |
+    | EEM 4         | SW0, MUXOUT0, MEZZ_IO8 |
+    | EEM 5         | SW1, MUXOUT1, MEZZ_IO11|
+    | EEM 6         | SW2, MUXOUT2, MEZZ_IO10|
+    | EEM 7         | SW3, MUXOUT3, MEZZ_IO13|
```
Comment (sbourdeauducq) [inline mirny.py:180]: Why change these?

File: mirny.py [hunk 7]
```diff
@@ -317,20 +317,52 @@ def __init__(self, platform):
...
+                self.comb += [ 
+                    tsi.o.eq(~regs[3].write[6]),
+                    regs[3].read[6].eq(tsi.i),
+                    tsi.oe.eq(regs[3].write[6 + 8]),
+                    regs[3].read[6 + 8].eq(tsi.oe),
+                ]
+
+            # hardcode SRCLR#
+            srclr = platform.request("almazny_srclr")
+            srclr.eq(1)
```
Comment (sbourdeauducq) [inline mirny.py:345]: No that doesn't work. Did you test this code?

File: mirny_cpld.py [hunk 8]
```diff
@@ -16,12 +16,17 @@
...
+        # IO from 0 to 7
+        ("mezz_io", 0, Pins("P57 P58 P59 P60 P61 P64 P68 P69")),
 
-        #almazny            CLK MOSI LT1 LT2  LT3 LT4 CLR
-        #                   IO8  11  10  13   15  0     3
-        ("mezz_io", 0, Pins("P94 P97 P96 P100 P92 P57 P60")),
-        #almazny                NOE1 NOE2 NOE3  4
+        # almazny                CLK MOSI LT1 LT2  LT3 LT4
+        #                       IO8  11  10  13   15  0   
+        ("almazny_io", 0, Pins("P94 P97 P96 P100 P92 P57")),
```
Comment (sbourdeauducq) [inline mirny_cpld.py:25]: There should be something that indicates that this is only for legacy Almazny.

File: mirny.py [hunk 9]
```diff
@@ -361,13 +397,4 @@ def __init__(self, platform):
...
-            ]
-
-        tp = [platform.request("tp", i) for i in range(5)]
-        self.comb += [
-            tp[0].eq(self.sr.ext.sck),
-            tp[1].eq(eem[1].i),
-            tp[2].eq(eem[2].i),
-            tp[3].eq(self.sr.bus.re),
-            tp[4].eq(self.sr.bus.we),
-        ]
```
Comment (sbourdeauducq) [inline mirny.py:373]: Why move this?

File: mirny_impl.py [hunk 10]
```diff
@@ -1,9 +1,21 @@
...
+    )
+    parser.add_argument("--legacy-almazny", action="store_true", default=False)
+
+    return parser
+
 def main():
     from mirny_cpld import Platform
     from mirny import Mirny
 
+    parsed = get_argparser().parse_args()
```
Comment (sbourdeauducq) [inline mirny_impl.py:15]: typically called ``args`` or similar elsewhere

File: mirny_cpld.py [hunk 11]
```diff
@@ -16,9 +16,18 @@
...
+        
+        # IO from 0 to 7
         ("mezz_io", 0, Pins("P57 P58 P59 P60 P61 P64 P68 P69")),
 
+        # almazny                CLK MOSI LT1 LT2  LT3 LT4
+        #                       IO8  11  10  13   15  0   
+        ("almazny_io", 0, Pins("P94 P97 P96 P100 P92 P57")),
+        # almazny                NOE1 NOE2 NOE3  4
+        ("almazny_noe", 0, Pins("P95 P98 P101 P58")),
+        # almazny SRCLR#
```
Comment (sbourdeauducq) [inline mirny_cpld.py:28]: This comment adds nothing to the code immediately below, remove.

File: mirny_cpld.py [hunk 12]
```diff
@@ -16,9 +16,18 @@
...
+        ("almazny_srclr", 0, Pins("P60")),
```
Comment (sbourdeauducq) [inline mirny_cpld.py:29]: should be ``nsrclr`` for consistency with ``noe``

File: mirny.py [hunk 13]
```diff
@@ -310,23 +310,59 @@ def __init__(self, platform):
...
-            self.comb += [
-                tsi.o.eq(regs[3].write[i] | (0 if i >= 4 else
-                    (regs[1].write[11] & eem[i + 4].i))),
-                regs[3].read[i].eq(tsi.i),
-                tsi.oe.eq(regs[3].write[i + 8]),
-                regs[3].read[i + 8].eq(tsi.oe),
-            ]
+        if legacy_almazny:
+            # 6 signals
+            for i, m in enumerate(platform.request("almazny_io")):
```
Comment (sbourdeauducq) [inline mirny.py:322]: Name them explicitly for consistency with ``noe`` and ``nsrclr``, don't just group them into one ``io`` vector.
You can still make a list of them and then use enumerate.

**Comment by jordens** [comment]
As mentioned above, I would highly recommend (1) using one of the reserved SPI prefixes and a 3-wire bus instead of bitbanging (also design the v1.1 interface w.r.t. that) and (2) just branch in git instead of extending the code to handle both.
And (3) I'd plan for the RF switches in v1.1 to be exposed directly and mirror the main RF switches.

**Comment by sbourdeauducq** [comment]
Replaced with #5
